php/elFinderVolumeGoogleDrive.class.php

```diff
@@ -0,0 +1,1659 @@
+					}
+					
+					$html = '<input id="elf-volumedriver-googledrive-host-btn" class="ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" value="{msg:btnApprove}" type="button" onclick="window.open(\''.$url.'\')">';
+					$html .= '<script>
+						$("#'.$options['id'].'").elfinder("instance").trigger("netmount", {protocol: "googledrive", mode: "makebtn"});
+					</script>';
+				}
+				
+				return array('exit' => true, 'body' => $html);
+			} else {
+					
+				$client = new Google_Client();
+																			   			
+    			$client->setAccessToken($this->session->get('elFinderGoogleDriveAuthTokens'));				
+				$oauth_token = $this->session->get('elFinderGoogleDriveAuthTokens');								
+				$this->session->set('elFinderGoogledriveTokens', array(rand(1000000000,9999999999), $oauth_token));
+				
+				$service = new Google_Service_Drive($client);
+				$folders = [];
+                    foreach($service->files->listFiles([
+                        'pageSize' => 1000,
+                        'q' => 'trashed = false and mimeType = "application/vnd.google-apps.folder"'
+                    ]) as $f) {
+                        $folders[$f->getId()] = $f->getName();
+                    }
+					
+				natcasesort($folders);
+				$folders = ['root' => 'My Drive'] + $folders;               
+				$folders = json_encode($folders);				
+				$options['pass'] = 'return';
+					
+				$out = array(
+					'node' => 'elfinder',
+					'json' => '{"protocol": "googledrive", "mode": "done", "folders": '.$folders.'}',
+					'bind' => 'netmount'
+				);
+				
+				return array('exit' => 'callback', 'out' => $out);
+			}
+
+		}
+		if ($this->session->get('elFinderGoogledriveTokens')) {
+		
+			list($options['googledriveUid'], $options['accessToken']) = $this->session->get('elFinderGoogledriveTokens');
+			$this->session->set('DriveCounter', (int)$this->session->get('DriveCounter')+1);
+				
+		}
+		unset($options['user'], $options['pass']);
+			
+		return $options;
+	}
+	
+	/**
+	 * process of on netunmount
+	 * Drop `googledrive` & rm thumbs
+	 * 
+	 * @param array $options
+	 * @return boolean
+	 */
+	public function netunmount($netVolumes, $key) {	
+		$count = 0;
+		$googledriveUid = '';
+		if (isset($netVolumes[$key])) {
+			$googledriveUid = $netVolumes[$key]['googledriveUid'];
+		}
+		foreach($netVolumes as $volume) {
+			if (@$volume['host'] === 'googledrive' && @$volume['googledriveUid'] === $googledriveUid) {
+				$count++;
+			}
+		}
+		if ($count === 1) {			
+			foreach(glob(rtrim($this->options['tmbPath'], '\\/').DIRECTORY_SEPARATOR.$this->tmbPrefix.'*.png') as $tmb) {
+				unlink($tmb);
+			}
+		}
+		if((int)$this->session->get('DriveCounter')>1){
+			$this->session->set('DriveCounter', (int)$this->session->get('DriveCounter')-1);
+		}else{	
+			$client = new Google_Client();
+																		   			
+			$client->setAccessToken($this->session->get('elFinderGoogleDriveAuthTokens'));									
+			$client->revokeToken();	
```

Comment:
revokeToken() is broken anothor fixed GoogleDrive volume drive with refreshToken. I think `$client->revokeToken()` is unnecessary.